### PR TITLE
add tailscale plugin

### PR DESCRIPTION
**Pull request by sheridans**
Initial tailscale plugin commit

needs the tailscale package

### CODE REVIEW

**Review by AdSchellevis — CHANGES_REQUESTED** [review]
@sheridans nice work, I've added some comments to improve the PR

File: security/tailscale/src/opnsense/mvc/app/controllers/OPNsense/Tailscale/Api/StatusController.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        if ($this->isJson($response)) {
+            return json_decode($response, true);
+        }
+
+    	return ['error' => 'Unable to determine Tailscale status, is the service running?'];
+    }
+
+    public function ipAction()
+    {
+        $response = trim((new Backend())->configdRun('tailscale tailscale-ip'));
```
[the remark this answers is not in the document]
Comment (sheridans) [reply security/tailscale/src/opnsense/mvc/app/controllers/OPNsense/Tailscale/Api/StatusController.php:52]: its plain text

File: security/tailscale/src/opnsense/mvc/app/controllers/OPNsense/Tailscale/Api/StatusController.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    	return ['result' => $response];
+    }
+
+    public function netAction()
+    {
+        $response = trim((new Backend())->configdRun('tailscale tailscale-netcheck'));
```
[the remark this answers is not in the document]
Comment (sheridans) [reply security/tailscale/src/opnsense/mvc/app/controllers/OPNsense/Tailscale/Api/StatusController.php:58]: also plain text response

File: security/tailscale/Makefile [hunk 3]
```diff
@@ -0,0 +1,8 @@
+PLUGIN_NAME=       tailscale
+PLUGIN_VERSION=    1.0
+PLUGIN_REVISION=   1
```
Comment (fichtner) [inline security/tailscale/Makefile:3]:
```suggestion
```

no revision yet

File: security/tailscale/Makefile [hunk 4]
```diff
@@ -0,0 +1,8 @@
+PLUGIN_NAME=       tailscale
+PLUGIN_VERSION=    1.0
```
Comment (fichtner) [inline security/tailscale/Makefile:2]:
```suggestion
PLUGIN_VERSION=    1.0.d
PLUGIN_DEVEL=        yes
```

just for the moment mark this as devel only

File: security/tailscale/pkg-descr [hunk 5]
```diff
@@ -0,0 +1,11 @@
+Tailscale is a mesh VPN alternative, based on WireGuard, that connects your
+computers, databases, and services together securely without any proxies.
+
+https://tailscale.com/
+
+Plugin Changelog
+================
+
+0.0.1
```
Comment (fichtner) [inline security/tailscale/pkg-descr:9]:
```suggestion
1.0
```

File: security/tailscale/LICENSE [hunk 6]
```diff
@@ -0,0 +1,21 @@
+BSD 2 Clause License
```
Comment (fichtner) [inline security/tailscale/LICENSE:1]: Don't really need this, your copyright gets folded into LICENSE page in root directory.. but from every page, not just this unattached one... and the BSD is standard amnyway :)

File: security/tailscale/Makefile (hunk 4 above)
Comment (sheridans) [reply security/tailscale/Makefile:2]: @fichtner thanks :)

File: security/tailscale/src/opnsense/mvc/app/controllers/OPNsense/Tailscale/Api/SettingsController.php [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    {
+        return $this->getBase("subnet4", "subnets.subnet4", $uuid);
+    }
+
+    public function delSubnetAction($uuid)
+    {
+        return $this->delBase("subnets.subnet4", $uuid);
+    }
+
+    public function reloadAction()
```
[the remark this answers is not in the document]
Comment (sheridans) [reply security/tailscale/src/opnsense/mvc/app/controllers/OPNsense/Tailscale/Api/SettingsController.php:62]: this was actually redundant, just needed removing

File: security/tailscale/src/opnsense/mvc/app/controllers/OPNsense/Tailscale/Api/StatusController.php [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+class StatusController extends ApiMutableModelControllerBase
+{
+    protected static $internalModelName = 'Tailscale';
+    protected static $internalModelClass = '\OPNsense\Tailscale\Status';
+
+    public function statusAction()
+    {
+        $response = trim((new Backend())->configdRun('tailscale tailscale-status'));
+
+        if ($this->isJson($response)) {
```
[the remark this answers is not in the document]
Comment (sheridans) [reply security/tailscale/src/opnsense/mvc/app/controllers/OPNsense/Tailscale/Api/StatusController.php]: Thanks for pointing out the obvious :)

File: security/tailscale/src/opnsense/service/templates/OPNsense/Tailscale/rc.conf.d [hunk 9]
```diff
@@ -0,0 +1,47 @@
+# DO NOT EDIT
+# THIS FILE IS AUTOMATICALLY GENERATED - ANY CHANGES WILL BE OVERWRITTEN
+#
+{% if helpers.exists('OPNsense.tailscale.settings.enabled') and OPNsense.tailscale.settings.enabled|default("0") == "1" %}
```
Comment (AdSchellevis) [inline security/tailscale/src/opnsense/service/templates/OPNsense/Tailscale/rc.conf.d:4]:
```suggestion
{% if not helpers.empty('OPNsense.tailscale.settings.enabled')  %}
```

a little bit less code, same applies to some of the others below

File: security/tailscale/src/opnsense/service/templates/OPNsense/Tailscale/rc.conf.d [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+{% if helpers.exists('OPNsense.tailscale.settings.listenPort') %}
+tailscaled_port="{{ OPNsense.tailscale.settings.listenPort }}"
+{% endif %}
+
+{% if helpers.exists('OPNsense.tailscale.settings.advertiseExitNode') and OPNsense.tailscale.settings.advertiseExitNode|default("0") == "1" %}
+tailscaled_exitnode_enable="YES"
+{% endif %}
+
+{% set up_args = "--reset" %}
```
Comment (AdSchellevis) [inline security/tailscale/src/opnsense/service/templates/OPNsense/Tailscale/rc.conf.d:15]:
```suggestion
{% set up_args = ["--reset"] %}
```

Let's convert this to a list, so we can stash all items and merge at the end.

File: security/tailscale/src/opnsense/service/templates/OPNsense/Tailscale/rc.conf.d [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+{% if helpers.exists('OPNsense.tailscale.settings.acceptSubnetRoutes') and OPNsense.tailscale.settings.acceptSubnetRoutes|default("0") == "1" %}
+{%   set up_args = up_args + " --accept-routes" %}
```
Comment (AdSchellevis) [inline security/tailscale/src/opnsense/service/templates/OPNsense/Tailscale/rc.conf.d:17]:
```suggestion
{%   do up_args.append(" --accept-routes") %}
```

File: security/tailscale/src/opnsense/service/templates/OPNsense/Tailscale/rc.conf.d [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+{% endif %}
+
+{% if helpers.exists('OPNsense.tailscale.settings.acceptDNS') and OPNsense.tailscale.settings.acceptDNS|default("0") == "1" %}
+{%   set up_args = up_args + " --accept-dns" %}
```
Comment (AdSchellevis) [inline security/tailscale/src/opnsense/service/templates/OPNsense/Tailscale/rc.conf.d:21]:
```suggestion
{%   do up_args.append(" --accept-dns") %}
```

File: security/tailscale/src/opnsense/service/templates/OPNsense/Tailscale/rc.conf.d [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+{% if helpers.exists('OPNsense.tailscale.authentication.preAuthKey') %}
+{%   set up_args = up_args + " --auth-key=" + OPNsense.tailscale.authentication.preAuthKey %}
+{% endif %}
+
+{# loop through subnets to build list #}
+{% if helpers.exists('OPNsense.tailscale.settings.subnets.subnet4') %}
+{%   set subnets = [] %}
+{%   for subnet_list in helpers.toList('OPNsense.tailscale.settings.subnets.subnet4') %}
+       {{ subnets.append(subnet_list.subnet) or "" }}
```
Comment (AdSchellevis) [inline security/tailscale/src/opnsense/service/templates/OPNsense/Tailscale/rc.conf.d:36]:
```suggestion
            {%   do subnets.append(subnet_list.subnet) %}
```

File: security/tailscale/src/opnsense/service/templates/OPNsense/Tailscale/rc.conf.d [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+{%   endfor %}
+{%   set subnetString = subnets|join(',') %}
+{%   set up_args = up_args + " --advertise-routes=" + subnetString %}
```
Comment (AdSchellevis) [inline security/tailscale/src/opnsense/service/templates/OPNsense/Tailscale/rc.conf.d:39]:
```suggestion
{%   do up_args.append(" --advertise-routes=" + subnetString) %}
```

File: security/tailscale/src/opnsense/service/templates/OPNsense/Tailscale/rc.conf.d [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+{% endif %}
+
+tailscaled_up_args="{{ up_args }}"
```
Comment (AdSchellevis) [inline security/tailscale/src/opnsense/service/templates/OPNsense/Tailscale/rc.conf.d:42]:
```suggestion
tailscaled_up_args="{{ up_args|join(' ') }}"
```

**Review by AdSchellevis — CHANGES_REQUESTED** [review]
@sheridans as promised some tips to simplify the template a bit further.

File: security/tailscale/src/opnsense/service/templates/OPNsense/Tailscale/rc.conf.d (hunk 9 above)
Comment (sheridans) [reply security/tailscale/src/opnsense/service/templates/OPNsense/Tailscale/rc.conf.d:4]: Awesome, thank you for that. Sorry, not familiar with jinja

File: security/tailscale/src/opnsense/mvc/app/views/OPNsense/Tailscale/authentication.volt [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+
+<div class="content-box">
+    {{ partial("layout_partials/base_form",['fields':authenticationForm,'id':'frmAuthentication']) }}
+</div>
+
+<section class="page-content-main">
+    <div class="content-box">
+        <div class="col-md-12">
+            <br/>
+            <div id="keaChangeMessage" class="alert alert-info" style="display: none" role="alert">
```
Comment (Monviech) [inline security/tailscale/src/opnsense/mvc/app/views/OPNsense/Tailscale/authentication.volt:29]: Probably copy pasta error in ID

File: security/tailscale/src/opnsense/mvc/app/views/OPNsense/Tailscale/settings.volt [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+    <li><a data-toggle="tab" href="#subnets" id="tab_subnets"> {{ lang._('Advertised Routes') }} </a></li>
+</ul>
+
+
+<div class="tab-content content-box">
+    <div id="settings"  class="tab-pane fade in active">
+        {{ partial("layout_partials/base_form",['fields':settingsForm,'id':'frmSettings'])}}
+    </div>
+    <div id="subnets" class="tab-pane fade in">
+        <table id="grid-subnets" class="table table-condensed table-hover table-striped" data-editDialog="DialogSubnet" data-editAlert="keaChangeMessage">
```
Comment (Monviech) [inline security/tailscale/src/opnsense/mvc/app/views/OPNsense/Tailscale/settings.volt:44]: Here too in editAlert

**Comment by sheridans** [comment]
All suggestions done, plus added missing acceptDNS field from settings form.

File: security/tailscale/src/opnsense/service/templates/OPNsense/Tailscale/rc.conf.d [hunk 18]
```diff
@@ -0,0 +1,46 @@
+# DO NOT EDIT
+# THIS FILE IS AUTOMATICALLY GENERATED - ANY CHANGES WILL BE OVERWRITTEN
+#
+{%  if not helpers.empty('OPNsense.tailscale.settings.enabled')  %}
+tailscaled_enable="YES"
```
Comment (fichtner) [inline security/tailscale/src/opnsense/service/templates/OPNsense/Tailscale/rc.conf.d:5]:
```suggestion
tailscaled_enable="YES"
tailscaled_env="TS_DEBUG_NETSTACK_SUBNETS=0"
```

can you test this one?

File: security/tailscale/src/etc/inc/plugins.inc.d/tailscale.inc [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+        'configd' => [
+            'restart' => ['tailscale restart'],
+            'start' => ['tailscale start'],
+            'stop' => ['tailscale stop'],
+        ],
+        'name' => 'tailscale',
+    ]; 
+
+    return $services;
+}
```
Comment (fichtner) [inline security/tailscale/src/etc/inc/plugins.inc.d/tailscale.inc:54]: let's see if this is enough

```suggestion
}

function tailscale_devices()
{
    return [[
        'pattern' => '^tailscale0$',
        'configurable' => false,
        'spoofmac' => false,
        'volatile' => true,
    ]];
}
```

File: security/tailscale/src/opnsense/service/templates/OPNsense/Tailscale/rc.conf.d (hunk 18 above)
Comment (sheridans) [reply security/tailscale/src/opnsense/service/templates/OPNsense/Tailscale/rc.conf.d:5]: with this enabled, can no longer access subnets behind opnsense from another tailscale device

File: security/tailscale/src/etc/inc/plugins.inc.d/tailscale.inc (hunk 19 above)
Comment (sheridans) [reply security/tailscale/src/etc/inc/plugins.inc.d/tailscale.inc:54]: This has now been added to PR

File: security/tailscale/src/opnsense/service/templates/OPNsense/Tailscale/rc.conf.d (hunk 18 above)
Comment (sheridans) [reply security/tailscale/src/opnsense/service/templates/OPNsense/Tailscale/rc.conf.d:5]: currently in template but commented out as it breaks, with a link to github thread

**Review by fichtner — APPROVED** [review]
ok for me

**Comment by sheridans** [comment]
Fine here

**Comment by fichtner** [comment]
Merged, thanks!

**Pull request merged**